warehouse/query-core/src/test/java/datawave/query/language/parser/jexl/JexlNodeSetTest.java

```diff
@@ -0,0 +1,266 @@
+    
+    @Test
+    public void testTheReasonForThisClass() {
+        JexlNode node1 = JexlNodeFactory.buildEQNode("FOO", "bar");
+        JexlNode node2 = JexlNodeFactory.buildEQNode("FOO", "bar");
+        
+        Set<JexlNode> nodeSet = Sets.newHashSet(node1, node2);
+        assertEquals(2, nodeSet.size());
+        assertTrue(nodeSet.contains(node1));
+        assertTrue(nodeSet.contains(node2));
+    }
+    
+    @Test
+    public void testGetNodesAndGetNodeKeys() {
+        JexlNode node = JexlNodeFactory.buildEQNode("FOO", "bar");
+        String key = "FOO == 'bar'";
+        
+        JexlNodeSet nodeSet = new JexlNodeSet();
+        nodeSet.add(node);
+        
+        assertEquals(Collections.singleton(key), nodeSet.getNodeKeys());
+        assertEquals(node, nodeSet.getNodes().iterator().next());
+    }
+    
+    @Test
+    public void testContains() {
+        JexlNode node1 = JexlNodeFactory.buildEQNode("FOO", "bar");
+        JexlNodeSet jexlNodeSet = new JexlNodeSet();
+        jexlNodeSet.add(node1);
+        assertTrue(jexlNodeSet.contains(node1));
+        
+        // Nope, it won't.
+        assertFalse(jexlNodeSet.contains("a string?!?!"));
+    }
+    
+    @Test
+    public void testIterator() {
+        JexlNode node = JexlNodeFactory.buildEQNode("FOO", "bar");
+        JexlNodeSet jexlNodeSet = new JexlNodeSet();
+        jexlNodeSet.add(node);
+        
+        Iterator<JexlNode> iter = jexlNodeSet.iterator();
+        assertTrue(iter.hasNext());
+        JexlNode next = iter.next();
+        assertEquals(JexlASTHelper.nodeToKey(node), JexlASTHelper.nodeToKey(next));
+        assertFalse(iter.hasNext());
+    }
+    
+    @Test
+    public void testToArray1() {
+        JexlNode node = JexlNodeFactory.buildEQNode("FOO", "bar");
+        JexlNodeSet nodeSet = new JexlNodeSet();
+        nodeSet.add(node);
+        
+        Object[] array = nodeSet.toArray();
+        assertEquals(1, array.length);
+    }
+    
+    @Test(expected = UnsupportedOperationException.class)
+    public void testToArray2() {
+        JexlNodeSet nodeSet = new JexlNodeSet();
+        nodeSet.toArray(new Object[0]);
+    }
+    
+    @Test
+    public void testAdd() {
+        JexlNode node1 = JexlNodeFactory.buildEQNode("FOO", "bar");
+        JexlNode node2 = JexlNodeFactory.buildEQNode("FOO", "bar");
+        JexlNode node3 = JexlNodeFactory.buildEQNode("FOO2", "bar2");
+        
+        JexlNodeSet nodeSet = new JexlNodeSet();
+        assertTrue(nodeSet.add(node1));
+        assertFalse(nodeSet.add(node2));
+        
+        assertEquals(1, nodeSet.size());
+        assertTrue(nodeSet.contains(node1));
+        assertTrue(nodeSet.contains(node2));
+        
+        assertTrue(nodeSet.add(node3));
+        assertFalse(nodeSet.add(node3));
+        assertFalse(nodeSet.add(node3));
+        
+        assertEquals(2, nodeSet.size());
+        assertTrue(nodeSet.contains(node1));
+        assertTrue(nodeSet.contains(node2));
```

Comment:
We never assert that `nodeSet.contains(node3)`.

Reply:
Will add